[DPE-6745] docs: add Kafka Connect tutorial (#328)
canonical/kafka-operator: docs/tutorial/t-kafka-connect.md

By the end, you should be able to use Kafka Connect integrator and Kafka Connect charms to streamline data ETL tasks on Canonical Data Platform charmed solutions.

## Prerequisites

We will be deploying different charmed data solutions including PostgreSQL and Opensearch. If you require more information or face issues deploying any of the mentioned products, you should consult the respective documentations:

- For PostgreSQL, refer to [Charmed PostgreSQL tutorial](https://charmhub.io/postgresql/docs/t-overview).
- For Opensearch, refer to [Charmed Opensearch tutorial](https://charmhub.io/opensearch/docs/tutorial).

## 0. Check current deployment

Up to this point, we should have a 3-unit Apache Kafka application, related to a 5-unit ZooKeeper application. That means the `juju status` command should show an output similar to the following:

> ```shell

Reviewer: **todo:** I *think* this should be `text`, here and further down.

Unit                         Workload  Agent  Machine  Public address  Ports     Message
data-integrator/0*           active    idle   8        10.38.169.159             
kafka/0                      active    idle   5        10.38.169.139             
kafka/1*                     active    idle   6        10.38.169.92              
kafka/2                      active    idle   7        10.38.169.70              
zookeeper/0*                 active    idle   0        10.38.169.164             
zookeeper/1                  active    idle   1        10.38.169.81              
zookeeper/2                  active    idle   2        10.38.169.72              
zookeeper/3                  active    idle   3        10.38.169.119             
zookeeper/4                  active    idle   4        10.38.169.215             
```

## 1. Set the necessary kernel properties for Opensearch

Since we will be deploying the Opensearch charm, we need to make necessary kernel configurations required for Opensearch charm to function properly, [described in detail here](https://charmhub.io/opensearch/docs/t-set-up#p-24545-set-kernel-parameters). This basically means running the following commands:

> ```shell

Reviewer: **nitpick:** `bash`, here and everywhere.

sudo tee -a /etc/sysctl.conf > /dev/null <<EOT
vm.max_map_count=262144
vm.swappiness=0
net.ipv4.tcp_retries2=5
fs.file-max=1048576
EOT

sudo sysctl -p
```

Next, we should set the required model parameters using the `juju model-config` command:

```shell
cat <<EOF > cloudinit-userdata.yaml
cloudinit-userdata: |
  postruncmd:
    - [ 'echo', 'vm.max_map_count=262144', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'vm.swappiness=0', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'net.ipv4.tcp_retries2=5', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'fs.file-max=1048576', '>>', '/etc/sysctl.conf' ]
    - [ 'sysctl', '-p' ]
EOF

juju model-config --file=./cloudinit-userdata.yaml
```

## 2. Deploy the databases and Kafka Connect charms

> Deploy the PostgreSQL, Opensearch, and Kafka Connect charms using the following commands. Since the Opensearch charm requires a TLS relation to become active, we will also need to deploy a TLS operator. We will be using the [`self-signed-certificates` charm](https://charmhub.io/self-signed-certificates) in this tutorial:

Reviewer: **note:** I think we cover TLS in the tutorial, so we could in theory assume it's existence here, as long as we put this tutorial towards the end of the the 'capital T' Tutorial.
We should update the `juju status` paste from earlier to include this.

Author: You're right, the tone isn't consistent. I'll wait for @izmalk's feedback on which tone is more appropriate for tutorials, and make changes accordingly.

data-integrator/0*           active    idle   8        10.38.169.159             
kafka-connect/0*             active    idle   9        10.38.169.23    8083/tcp  
kafka/0                      active    idle   5        10.38.169.139             
kafka/1*                     active    idle   6        10.38.169.92              
kafka/2                      active    idle   7        10.38.169.70              
opensearch/0*                active    idle   10       10.38.169.172   9200/tcp  
postgresql/0*                active    idle   12       10.38.169.121   5432/tcp  Primary
self-signed-certificates/0*  active    idle   11       10.38.169.82              
zookeeper/0*                 active    idle   0        10.38.169.164             
zookeeper/1                  active    idle   1        10.38.169.81              
zookeeper/2                  active    idle   2        10.38.169.72              
zookeeper/3                  active    idle   3        10.38.169.119             
zookeeper/4                  active    idle   4        10.38.169.215             
```

## 4. Load some test data into the PostgreSQL database

> In a real-world scenario, you will have some application writing data in PostgreSQL database(s). However, for the sake of this tutorial, we will be generating some test data using the SQL script provided in [this GitHub gist](https://gist.github.com/imanenami/bc5900ed2b58e0cc980b498d04677095) and load it into a PostgreSQL database using the `psql` bin command shipped with the PostgreSQL charm.

Reviewer: **todo:** IMO, the Gist is small, let's just put it here and give the full instructions. At least that way it exists here full time.

Author: Agree, Done :)

[Note]
For more information on how to access a PostgreSQL database using the PostgreSQL charm, refer to [Access PostgreSQL](https://charmhub.io/postgresql/docs/t-access) section of the Charmed PostgreSQL tutorial.
[/Note]

First, download the SQL script from the GitHub gist:

```shell

> wget https://gist.github.com/imanenami/bc5900ed2b58e0cc980b498d04677095/raw/4169eb100bd65de9264847c822202cd204516aac/populate.sql -O ./populate.sql

Reviewer: **todo:** Small request to set file to `/tmp`

```

Next, copy the `populate.sql` script to the PostgreSQL unit using the `juju scp` command:

```shell
juju scp populate.sql postgresql/0:/home/ubuntu/populate.sql
```

Then, following the [Access PostgreSQL](https://charmhub.io/postgresql/docs/t-access) tutorial, grab the `operator` user's password on PostgreSQL database using the `get-password` action:

```shell

> juju run postgresql/leader get-password

Reviewer: **todo:** Based on above comment, we could just `<< EOL ; EOL` here, that way they can still copy the code-block that this renders to.

```

Sample output:

> ```
> ...
> password: bQOUgw8ZZgUyPA6n
> ```

Reviewer: **todo:**
````suggestion
```text
...
password: bQOUgw8ZZgUyPA6n
```
````

Make note of the password, and ssh into the PostgreSQL unit:

> ```
> juju ssh postgresql/leader
> ```

Reviewer: **todo:**
````suggestion
```bash
juju ssh postgresql/leader
```
````

Once inside the virtual machine, you can use the `psql` command line interface using `operator` user credentials, to create the `tutorial` database:

```shell
psql --host <postgresql-unit-ip> --username operator --password --dbname postgres \
    -c "CREATE DATABASE tutorial"
```

You will be prompted to type the password, which you have obtained previously.

> Now, we could use the `populate.sql` script copied earlier into the PostgreSQL unit, to create a table named `posts` with some test data:

Reviewer: **todo:** I noticed here that we switch between using `we` pronouns and `you` pronouns. I don't have a preference, but let's try to be consistent across the whole of this tutorial doc and just pick one. Whatever the rest of the tutorial uses more should be OK.

Author: Thanks, added: https://github.com/canonical/kafka-operator/pull/328/commits/a90e21a90097dc8aba7deacc8cfa05df34d70ed4

Author: Good suggestion, done here: https://github.com/canonical/kafka-operator/pull/328/commits/c29538b03df21d079d1354333a621bc2e18a4763

```shell
cat /home/ubuntu/populate.sql | \
    psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial
```

To ensure that the test data is loaded successfully into the `posts` table, use the following command:

```shell
psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial \
    -c 'SELECT COUNT(*) FROM posts'
```

Which should have an output like the following, indicating that 5 rows have been added to the `posts` table: 

> ```shell
>  count 
> -------
>      5
> (1 row)
> ```

Reviewer: **todo:**
````suggestion
```text
 count 
-------
     5
(1 row)
```
````

Log out from the PostgreSQL unit using `exit` command or the `Ctrl+D` keyboard shortcut.

## 5. Deploy and integrate the `postgresql-connect-integrator` charm

> Now that you have some data loaded into PostgreSQL, it is time to deploy the `postgresql-connect-integrator` charm to enable integration of PostgreSQL and Kafka Connect applications. First, deploy the charm using `juju deploy` command and provide the minimum necessary configurations:

Reviewer: **suggestion:** You mentioned `sink` mode later on with OS, I think we should briefly mention 'deploying this in `source` mode' here.

```shell
juju deploy postgresql-connect-integrator \
    --channel edge \
    --config mode=source \
    --config db_name=tutorial \
    --config topic_prefix=etl_
```

Each Kafka Connect integrator application needs at least two relations to activate: one relation with the Kafka Connect application and one or more relations with Data charms (e.g. MySQL, PostgreSQL, Opensearch, etc.) depending on the use-case.

Therefore, in order to activate this particular integrator charm, use the `juju integrate` command to integrate it with the Kafka Connect and PostgreSQL applications:

```shell
juju integrate postgresql-connect-integrator postgresql
juju integrate postgresql-connect-integrator kafka-connect
```

After a couple of minutes, `juju status` command should show the `postgresql-connect-integrator` in `active|idle` state, with a message indicating that the ETL task is running:

> ```
> ...
> postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
> ...
> ```

Reviewer: **todo:**
````suggestion
```text
...
postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
...
```
````

## 6. Deploy and integrate the `opensearch-connect-integrator` charm

You are almost done with the ETL task, the only remaining part is to move data from Apache Kafka to Opensearch. To achieve that, you need to deploy another Kafka Connect integrator named `opensearch-connect-integrator` in `sink` mode:

```shell
juju deploy opensearch-connect-integrator \
    --channel edge \
    --config mode=sink \
    --config topics="etl_posts"
```

The above command would deploy an integrator application to move messages published into the `etl_posts` topic to an index in Opensearch named `etl_posts`. We know that the `etl_posts` topic is being filled by the `postgresql-connect-integrator` charm we deployed in step 5.

To activate the `opensearch-connect-integrator`, make the necessary integrations:

```shell
juju integrate opensearch-connect-integrator opensearch
juju integrate opensearch-connect-integrator kafka-connect
```

Wait a couple of minutes and run `juju status`, now both `opensearch-connect-integrator` and `postgresql-connect-integrator` applications should be in `active|idle` state, showing a message indicating that the ETL task is running:

> ```
> ...
> opensearch-connect-integrator/0*  active    idle   14       10.38.169.108   8080/tcp  Task Status: RUNNING
> ...
> postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
> ...
> ```

Reviewer: **todo:**
````suggestion
```text
...
opensearch-connect-integrator/0*  active    idle   14       10.38.169.108   8080/tcp  Task Status: RUNNING
postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
...
```
````

## 7. Verify the data is being copied

> Now it's time to verify that the data is being copied from the PostgreSQL database to the Opensearch index. We could use the Opensearch REST API for that purpose.

Reviewer: **very nitpicky suggestion:** I've seen this come up a bit, but the phrasing `We could use the...` sounds sort of like saying 'We could do blahblah, but maybe not'. Changing it to `We can use the...` sounds to me like 'Yeah! WE CAN do blahblah'. Just a note in case it comes up again.

Author: Great one, really appreciate these types of suggestions from you ;)

First, retrieve the admin user credentials for Opensearch using `get-password` action:

```shell
juju run opensearch/leader get-password
```

Sample output:

> ```
> ...
> password: GoCNE5KdFywT4nF1GSrwpAGyqRLecSXC
> username: admin
> ```

Reviewer: **todo:**
````suggestion
```text
...
password: GoCNE5KdFywT4nF1GSrwpAGyqRLecSXC
username: admin
```
````

Now, using the password obtained above, send a request to the topic's `_search` endpoint, either using your browser or `curl`:

```shell
curl -u admin:<admin-password> -k -X GET https://<opensearch-unit-ip>:9200/etl_posts/_search
```

You will get a JSON response containing the search results, which should have 5 documents. A truncated sample output is shown below (Note the `hits.total` value which should be 5):

> ```
> {
>   "took": 15,
>   "timed_out": false, 
>   "_shards": {
>     "total": 1,
>     "successful": 1,
>     "skipped": 0,
>     "failed": 0
>   },
>   "hits": {
>     "total": {
>       "value": 5,
>       "relation": "eq"
>     },
>     "max_score": 1.0,
>     "hits": [
>       ...
>     ]
>   }
> }
> 
> ```

Reviewer: **todo:**
````suggestion
```text
{
  "took": 15,
  "timed_out": false, 
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 5,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      ...
    ]
  }
}

```
````

> Now let's insert a new post into the PostgreSQL database. First ssh into the PostgreSQL leader unit:

Reviewer: **nitpick:**
```suggestion
Now let's insert a new post into the PostgreSQL database. First SSH in to the PostgreSQL leader unit:
```

Reviewer: @a-velasco - Do we have to be careful with PostgreSQL/Opensearch terminology here? We have to refer to Kafka + ZooKeeper as 'Apache' Foo, any brand requirements here we need to be aware of?

Reviewer 2: At the moment, there are no such requirements for PostgreSQL, MySQL, OpenSearch, or MongoDB.

```shell
juju ssh postgresql/leader
```

Then, insert a new post using following command and the password you obtained in step 4:

```shell
psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial -c \ 
    "INSERT INTO posts (content, likes) VALUES ('my new post', 1)"
```

Log out from the PostgreSQL unit using `exit` command or the `Ctrl+D` keyboard shortcut.

Then, check that the data is automatically copied to the Opensearch index:

```shell
curl -u admin:<admin-password> -k -X GET https://<opensearch-unit-ip>:9200/etl_posts/_search
```

Which now should have 6 hits (output is truncated):

> ```
> {
> ...
>   "hits": {
>     "total": {
>       "value": 6,
>       "relation": "eq"
>     },
> ...
> }
> ```

Reviewer: **todo:** I'm so sorry, this is the last one.
````suggestion
```text
{
...
  "hits": {
    "total": {
      "value": 6,
      "relation": "eq"
    },
...
}
```
````

# Move Data Between Data Platform Charms using Kafka Connect

In this part of the tutorial, we are going to use [Kafka Connect](https://kafka.apache.org/documentation/#connect) - an ETL framework on top of Apache Kafka - to seamlessly move data between different charmed database technologies.

> We will follow a step-by-step process for moving data between [Canonical Data Platform charms](https://canonical.com/data) using Kafka Connect. Specifically, we will showcase a particular use-case of loading data from a relational database, i.e. PostgreSQL, to a document store and search engine, i.e. Opensearch, entirely using charmed solutions.

Reviewer 2: ```suggestion
We will follow a step-by-step process for moving data between [Canonical Data Platform charms](https://canonical.com/data) using Kafka Connect. Specifically, we will showcase a particular use-case of loading data from a relational database, i.e. PostgreSQL, to a document store and search engine, i.e. OpenSearch, entirely using charmed solutions.
```
**Nitpick:** Could be worth doing a quick search & replace to change `Opensearch` into `OpenSearch` throughout the page -- 1. for "correct" capitalization and 2. to avoid work in the future when we implement some doc workflows that check for this kind of stuff

By the end, you should be able to use Kafka Connect integrator and Kafka Connect charms to streamline data ETL tasks on Canonical Data Platform charmed solutions.

## Prerequisites

> We will be deploying different charmed data solutions including PostgreSQL and Opensearch. If you require more information or face issues deploying any of the mentioned products, you should consult the respective documentations:

Reviewer 2: I think it would be more helpful in this section to summarize the prerequisites in terms of 
* what should I already have deployed that this tutorial won't cover (e.g. 3 units of kafka, 5 units zookeper)
* what hardware requirements I should fulfill to handle what we'll be deploying (namely memory and disk space)

Since you explicitly include the deployment of PostgreSQL and OpenSearch in the tutorial, I'd suggest not mentioning them here to avoid ambiguity around whether they should be deployed already. Instead, these tutorial links could be added to section "2. Deploy the databases and Kafka Connect charms" as resources for additional information.

kafka                     3.6.1    active      3  kafka                     3/stable       195  no       
zookeeper                 3.8.4    active      5  zookeeper                 3/stable       149  no       

Unit                         Workload  Agent  Machine  Public address  Ports     Message
data-integrator/0*           active    idle   8        10.38.169.159             
kafka/0                      active    idle   5        10.38.169.139             
kafka/1*                     active    idle   6        10.38.169.92              
kafka/2                      active    idle   7        10.38.169.70
self-signed-certificates/0*  active    idle   9        10.38.169.82                            
zookeeper/0*                 active    idle   0        10.38.169.164             
zookeeper/1                  active    idle   1        10.38.169.81              
zookeeper/2                  active    idle   2        10.38.169.72              
zookeeper/3                  active    idle   3        10.38.169.119             
zookeeper/4                  active    idle   4        10.38.169.215             
```

> ## 1. Set the necessary kernel properties for Opensearch

Reviewer 2: I like to number instructions when the chronological order of the events is particularly important and I want to highlight it. That being said, I agree that on h2 headings it doesn't look very good.

Since this is a tutorial, I think it's implied that headings are to be followed in order, so I think it's pretty safe to remove the numbering.

> ## 1. Set the necessary kernel properties for Opensearch

Author: Agreed, Let's remove the numbers. Thanks!

> Since we will be deploying the Opensearch charm, we need to make necessary kernel configurations required for Opensearch charm to function properly, [described in detail here](https://charmhub.io/opensearch/docs/t-set-up#p-24545-set-kernel-parameters). This basically means running the following commands:

Author: Personally, I don't like to jump between tutorials, so I thought including the commands here and putting a reference link to the original tutorial for readers interested to read more makes sense.

vm.swappiness=0
net.ipv4.tcp_retries2=5
fs.file-max=1048576
EOT

sudo sysctl -p
```

Next, we should set the required model parameters using the `juju model-config` command:

```bash
cat <<EOF > cloudinit-userdata.yaml
cloudinit-userdata: |
  postruncmd:
    - [ 'echo', 'vm.max_map_count=262144', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'vm.swappiness=0', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'net.ipv4.tcp_retries2=5', '>>', '/etc/sysctl.conf' ]
    - [ 'echo', 'fs.file-max=1048576', '>>', '/etc/sysctl.conf' ]
    - [ 'sysctl', '-p' ]
EOF

juju model-config --file=./cloudinit-userdata.yaml
```

## 2. Deploy the databases and Kafka Connect charms

Deploy the PostgreSQL, Opensearch, and Kafka Connect charms using the following commands. Since the Opensearch charm requires a TLS relation to become active, we will also need a TLS operator, for which We will be using the [`self-signed-certificates` charm](https://charmhub.io/self-signed-certificates) that was deployed in the [Enable Encryption](https://charmhub.io/kafka/docs/t-enable-encryption) part of the Charmed Kafka's Tutorial.

```bash
juju deploy kafka-connect --channel edge
juju deploy postgresql --channel 14/stable

> juju deploy opensearch --channel 2/edge --config profile=testing

Reviewer 2: FYI: OpenSearch can now be deployed from the stable channel.

If it needs to be deployed from edge for a specific reason, ignore this!

Author: Great, thanks! fixed here: https://github.com/canonical/kafka-operator/pull/328/commits/aca340ce66668a641e45b892579309dfe32537f7

kafka/1*                     active    idle   6        10.38.169.92              
kafka/2                      active    idle   7        10.38.169.70              
opensearch/0*                active    idle   11       10.38.169.172   9200/tcp  
postgresql/0*                active    idle   12       10.38.169.121   5432/tcp  Primary
self-signed-certificates/0*  active    idle   9       10.38.169.82              
zookeeper/0*                 active    idle   0        10.38.169.164             
zookeeper/1                  active    idle   1        10.38.169.81              
zookeeper/2                  active    idle   2        10.38.169.72              
zookeeper/3                  active    idle   3        10.38.169.119             
zookeeper/4                  active    idle   4        10.38.169.215             
```

## 4. Load some test data into the PostgreSQL database

In a real-world scenario, you will have some application writing data in PostgreSQL database(s). However, for the sake of this tutorial, we will be generating some test data using a simple SQL script and load it into a PostgreSQL database using the `psql` bin command shipped with the PostgreSQL charm.

> [Note]
> For more information on how to access a PostgreSQL database using the PostgreSQL charm, refer to [Access PostgreSQL](https://charmhub.io/postgresql/docs/t-access) section of the Charmed PostgreSQL tutorial.
> [/Note]

Reviewer 2: ```suggestion
[note]
For more information on how to access a PostgreSQL database using the PostgreSQL charm, refer to [Access PostgreSQL](https://charmhub.io/postgresql/docs/t-access) section of the Charmed PostgreSQL tutorial.
[/note]
```

I don't think capitalization would cause issues with discourse markdown, but just in case.

> First, create the SQL script in the `/tmp/populate.sql` path, using the following command:

Author: Thanks, applied here with slight mod: https://github.com/canonical/kafka-operator/pull/328/commits/51bd28d9c9e549e697407d8ba0e8682bbf822787

```bash

> cat <<EOF > /tmp/populate.sql

Author: The /tmp addition was from Marc's suggestion, and I think it's good too. Plus, I don't think an absolute system root path `/something` is a good idea, so skipping this. Thanks!

    1000
  ), 
  (
    'Charmed Apache Kafka is developed and supported by Canonical, as part of its commitment to provide open-source, self-driving solutions, seamlessly integrated using the Operator Framework Juju. Please refer to Charmhub, for more charmed operators that can be integrated by Juju.', 
    60
  );
EOF
```

Next, copy the `populate.sql` script to the PostgreSQL unit using the `juju scp` command:

```bash
juju scp /tmp/populate.sql postgresql/0:/home/ubuntu/populate.sql
```

Then, following the [Access PostgreSQL](https://charmhub.io/postgresql/docs/t-access) tutorial, grab the `operator` user's password on PostgreSQL database using the `get-password` action:

```bash
juju run postgresql/leader get-password
```

Sample output:

```text
...
password: bQOUgw8ZZgUyPA6n
```

Make note of the password, and ssh into the PostgreSQL unit:

```bash
juju ssh postgresql/leader
```

Once inside the virtual machine, you can use the `psql` command line interface using `operator` user credentials, to create the `tutorial` database:

```bash

> psql --host <postgresql-unit-ip> --username operator --password --dbname postgres \

Author: We could use `hostname -i` because we disable ipv6 at the setup env phase. Replaced here: https://github.com/canonical/kafka-operator/pull/328/commits/fed1ce7c31c4d4bd3bd35f1cdc96cf69509f1230

Thanks!

    -c "CREATE DATABASE tutorial"
```

You will be prompted to type the password, which you have obtained previously.

Now, we can use the `populate.sql` script copied earlier into the PostgreSQL unit, to create a table named `posts` with some test data:

```bash
cat /home/ubuntu/populate.sql | \
    psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial
```

To ensure that the test data is loaded successfully into the `posts` table, use the following command:

```bash
psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial \
    -c 'SELECT COUNT(*) FROM posts'
```

Which should have an output like the following, indicating that 5 rows have been added to the `posts` table: 

```text
 count 
-------
     5
(1 row)
```

Log out from the PostgreSQL unit using `exit` command or the `Ctrl+D` keyboard shortcut.

> ## 5. Deploy and integrate the `postgresql-connect-integrator` charm

Author: The second suggestion is ambiguous because the reader might think we mean the postgres charm itself. So I went with the first suggestion

    --config db_name=tutorial \
    --config topic_prefix=etl_
```

Each Kafka Connect integrator application needs at least two relations to activate: one relation with the Kafka Connect application and one or more relations with Data charms (e.g. MySQL, PostgreSQL, Opensearch, etc.) depending on the use-case.

Therefore, in order to activate this particular integrator charm, use the `juju integrate` command to integrate it with the Kafka Connect and PostgreSQL applications:

```bash
juju integrate postgresql-connect-integrator postgresql
juju integrate postgresql-connect-integrator kafka-connect
```

After a couple of minutes, `juju status` command should show the `postgresql-connect-integrator` in `active|idle` state, with a message indicating that the ETL task is running:

```text
...
postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
...
```

This means that the integrator application is actively copying data from the source database, i.e. `tutorial` into Apache Kafka topics prefixed with `etl_`. For example, rows in the `posts` table will be published into an Apache Kafka topic named `etl_posts`.

> ## 6. Deploy and integrate the `opensearch-connect-integrator` charm

Author: Skipping this suggestion because `OpenSearch charm` is ambiguous, I remove the number in another commit :)

> You are almost done with the ETL task, the only remaining part is to move data from Apache Kafka to Opensearch. To achieve that, you need to deploy another Kafka Connect integrator named `opensearch-connect-integrator` in `sink` mode:

Reviewer 3: Suggestion: Split the sentences for maintenance.

```suggestion
You are almost done with the ETL task, the only remaining part is to move data from Apache Kafka to Opensearch. 
To do that, deploy another Kafka Connect integrator named `opensearch-connect-integrator` in the `sink` mode:
```

```bash
juju deploy opensearch-connect-integrator \
    --channel edge \
    --config mode=sink \
    --config topics="etl_posts"
```

> The above command would deploy an integrator application to move messages published into the `etl_posts` topic to an index in Opensearch named `etl_posts`. We know that the `etl_posts` topic is being filled by the `postgresql-connect-integrator` charm we deployed in step 5.

Reviewer 3: Suggestions: wording improvements and sentence split.

```suggestion
The above command deploys an integrator application to move messages from the `etl_posts` topic to the index in OpenSearch named `etl_posts`. 
And the `etl_posts` topic is filled by the `postgresql-connect-integrator` charm we deployed earlier.
```

To activate the `opensearch-connect-integrator`, make the necessary integrations:

```bash
juju integrate opensearch-connect-integrator opensearch
juju integrate opensearch-connect-integrator kafka-connect
```

Wait a couple of minutes and run `juju status`, now both `opensearch-connect-integrator` and `postgresql-connect-integrator` applications should be in `active|idle` state, showing a message indicating that the ETL task is running:

```text
...
opensearch-connect-integrator/0*  active    idle   14       10.38.169.108   8080/tcp  Task Status: RUNNING
postgresql-connect-integrator/0*  active    idle   13       10.38.169.83    8080/tcp  Task Status: RUNNING
...
```

> ## 7. Verify the data is being copied

Reviewer 3: Suggestion: to make the heading shorter and remove enumeration.

```suggestion
## Verify data transfer
```

> Now it's time to verify that the data is being copied from the PostgreSQL database to the Opensearch index. We can use the Opensearch REST API for that purpose.

Reviewer 3: Suggestion: Sentence split and OpenSearch naming capitalision fix.
```suggestion
Now it's time to verify that the data is being copied from the PostgreSQL database to the OpenSearch index. 
We can use the OpenSearch REST API for that purpose.
```

> First, retrieve the admin user credentials for Opensearch using `get-password` action:

Reviewer 3: ```suggestion
First, retrieve the admin user credentials for OpenSearch using `get-password` action:
```

```bash
juju run opensearch/leader get-password
```

> Sample output:

Reviewer 3: Suggestion: Wording improvements and add an empty line before a code block.

```suggestion
As a result, you should see output similar to the following:

```

```text
...
password: GoCNE5KdFywT4nF1GSrwpAGyqRLecSXC
username: admin
```

Now, using the password obtained above, send a request to the topic's `_search` endpoint, either using your browser or `curl`:

```bash

> curl -u admin:<admin-password> -k -X GET https://<opensearch-unit-ip>:9200/etl_posts/_search

Reviewer 3: Suggestion: Let's add direct instructions on how to get the IP address needed.

Author: Good tip, done here: https://github.com/canonical/kafka-operator/pull/328/commits/59194852a652f6c9264e1a4d39a2f8e06e770510

```

> You will get a JSON response containing the search results, which should have 5 documents. A truncated sample output is shown below (Note the `hits.total` value which should be 5):

Reviewer 3: Suggestion: Wording improvements for clarity. Also, we should try to avoid using digits like 1-9 in the usual text. Ideally, replace them with words or literal values (in backticks).

```suggestion
As a result you get a JSON response containing the search results, which should have five documents. 
The `hits.total` value should be `5`, as shown in the output example below:
```

```text
{
  "took": 15,
  "timed_out": false, 
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 5,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      ...
    ]
  }
}

```

Now let's insert a new post into the PostgreSQL database. First SSH in to the PostgreSQL leader unit:

```bash
juju ssh postgresql/leader
```

> Then, insert a new post using following command and the password you obtained in step 4:

Reviewer 3: Suggestion: Improve wording.

```suggestion
Then, insert a new post using following command and the password for the `operator` user on the PostgreSQL:
```

```bash
psql --host <postgresql-unit-ip> --username operator --password --dbname tutorial -c \ 
    "INSERT INTO posts (content, likes) VALUES ('my new post', 1)"
```

Log out from the PostgreSQL unit using `exit` command or the `Ctrl+D` keyboard shortcut.

> Then, check that the data is automatically copied to the Opensearch index:

Reviewer 3: Suggestion: OpenSearch naming.

```suggestion
Then, check that the data is automatically copied to the OpenSearch index:
```

```bash
curl -u admin:<admin-password> -k -X GET https://<opensearch-unit-ip>:9200/etl_posts/_search
```

> Which now should have 6 hits (output is truncated):

Reviewer 3: Suggestion: replace a single digit number with a word.

```suggestion
Which now should have six hits (output is truncated):
```

```text
{
...
  "hits": {
    "total": {
      "value": 6,
      "relation": "eq"
    },
...
}
```

> Congratualtions! You have successfully deployed an ETL job to move data continuously from PostgreSQL to Opensearch, entirely using charmed solutions.

Reviewer 3: Suggestion: Wording improvements.

```suggestion
Congratulations! You have successfully completed an ETL job that continuously moves data from PostgreSQL to OpenSearch, using entirely charmed solutions.
```
